doc/man3/SSL_CTX_set1_curves.pod

```diff
@@ -66,10 +66,11 @@ client's order determines the selection.
 selection is more complex. A TLS 1.3 client sends both a group list as well as a
 predicted subset of groups. Choosing a group outside the predicted subset incurs
 an extra roundtrip. However, in some situations, the most preferred group may
-not be predicted. OpenSSL considers all supported groups to be equal in security
-and prioritizes avoiding roundtrips. If an application uses an external provider
-to extend OpenSSL with, e.g., a post-quantum algorithm, this behavior may allow
-a network attacker to downgrade connections to a weaker algorithm.
+not be predicted. OpenSSL considers all supported groups to be comparable in
+security and prioritizes avoiding roundtrips above either client or server
+preference order. If an application uses an external provider to extend OpenSSL
+with, e.g., a post-quantum algorithm, this behavior may allow a network attacker
+to downgrade connections to a weaker algorithm.
```

Comment:
Can you explain how this attack would work?

Reply:
Sure. This was also discussed in the original bug, the IETF discussion, and the Internet-Draft, but I can describe it anew here.

Concerns like downgrade and whatnot are about the behavior of the client and server combined, so we need to look at several things and what meaning the protocol specification has ascribed to each message:

1. What is OpenSSL's response to ClientHellos like...
   * key_shares=[unrecognized_future_pq2, X25519], supported_groups=[unrecognized_future_pq2, ML-KEM, X25519]
   * key_shares=[X25519], supported_groups=[ML-KEM, X25519]
2. What does the client mean when it sends such a ClientHello, per the specification?
3. Under what circumstances might a client reasonably do so, and where do attackers come into play?

Taking these in order:

OpenSSL will respond to those ClientHellos by picking X25519 instead of ML-KEM. This is because OpenSSL prioritizes roundtrips without looking at the client's supported_groups extension at all. That is, OpenSSL assumes all supported groups to be comparable enough in security that the cost of a round-trip outweighs it. I think this assumption is clearly wrong when the groups are X25519 vs ML-KEM, during a post-quantum transition.

Now, what do this messages actually mean? As I said, downgrade depends on both behaviors, with the RFC mediating the semantics of each message, so whether this is a problem or a misbehaving client depends on the spec. RFC 8446 is quite clear here:

> Each KeyShareEntry value MUST correspond to a
>   group offered in the "supported_groups" extension and MUST appear in
>   the same order.  However, the values MAY be a non-contiguous subset
>   of the "supported_groups" extension and MAY omit the most preferred
>   groups.  Such a situation could arise if the most preferred groups
>   are new and unlikely to be supported in enough places to make
>   pregenerating key shares for them efficient.

A server may _not_ assume omitted key shares were considered undesirable by the client. They have been 
omitted for other reasons. rfc8446bis will further clarify this, added in light of discovering OpenSSL's post-quantum-incompatibility:

> For this reason, the omission of a share for group A and inclusion of one for group B does not mean that the client prefers B to A. Selecting a group based on KeyShareEntry may result in the use of a less preferred group than the client and server mutually support, though saving the round trip of HelloRetryRequest. Servers that wish to respect the client's group preferences SHOULD first select a group based "supported_groups" and then either send a ServerHello or a HelloRetryRequest depending on the contents of KeyshareClienthello.

Of course, that last bit is only for servers that wish to respect client preferences. TLS ultimately leaves the selection algorithm to the implementation. But if OpenSSL picks a different algorithm, we need to be able to articulate the semantics and implications. In this case, the OpenSSL algorithm is to ignore _both_ the client's preferences _and_ the caller's preference order and use a avoid-round-trips-first policy, independent of security level.

Finally, what are the implications? Why would a client send such a thing, and where do attackers come into play? Here, once again, the concerns fall directly out of post-quantum. Post-quantum KEMs are huge. This causes problems today, and problems in the future.

Let's imagine, later this year, some OpenSSL user sees the post-quantum advertisement in the changelog, [blogs](https://www.openssl.org/blog/blog/2023/09/07/ossl32a1/index.html), etc., and installed a post-quantum provider with ML-KEM. They read OpenSSL's current documentation and say, ok, I'll put ML-KEM before X25519.

First, the problems today: some clients are already finding compatibility impacts of predicting the PQ KEM are high enough that they have to resort to heuristics where they don't offer the PQ KEM sometimes. Chrome, thankfully, has not needed this yet, but [Cloudflare](https://mailarchive.ietf.org/arch/msg/tls/X8roz2X5kmekEULrHa_xbzp3TkU/) has. When those heuristics fire, either under attacker control or simply unconditionally, OpenSSL will pick X25519, a downgrade from what this OpenSSL user reasonably expected.

Now the problems in the future. Even without compat implications, clients cannot reasonably send two PQ KEMs in the initial ClientHello. That would be too expensive. Now, that's fine today, assuming everyone agrees on an ML-KEM-768 / X25519 hybrid. But we can't assume that will be the last KEM. Our hypothetical OpenSSL user values stability, so they would like to know their server deployments will work into the future.

Sometime later, we may develop a better PQ KEM, PQ2. Maybe it's faster or smaller. Clients start to transition. Again, we cannot send both, so the plausible paths are:
* Clients start to use hints like DNS to decide which key share to send
* Absent a DNS hint, clients start to offer PQ2 instead of ML-KEM in key shares, still listing both in ML-KEM
* Or perhaps the client uses some other heuristic

All of these size mitigations are supported by the RFC. The RFC is very clear that clients are permitted to omit predicted key shares. But, once again, OpenSSL's behavior will pick a weaker algorithm when the mitigations misfire. Misfirings may be under attacker influence (DNS is often unauthenticated), or just bad luck. Either way, we do not get the secure settings the OpenSSL user wanted.

Hopefully that clears things up, and makes it more clearer how OpenSSL's current behavior is not post-quantum-ready.

Reply:
Thanks for the writeup. This is valuable information.

My direct question is: assuming both endpoints are configured so that a PQC KEM *is* negotiated, is there a viable downgrade attack by an on-path active MitM attacker here?

Is there an *attack* here or is this simply a suboptimal outcome?

Reply:
I don't understand the question. It doesn't make much sense given the context.

> My direct question is: assuming both endpoints are configured so that a PQC KEM is negotiated, is there a viable downgrade attack by an on-path active MitM attacker here?

Well, as discussed above, it is *not possible* to configure OpenSSL such that a PQC KEM is reliably negotiated when you also support classical algorithms. OpenSSL's claim in CHANGES.md, and the documentation, prior to this PR, reads as if it is possible, but such a configuration does not do what your documentation claims.

That means, yes, there is a viable downgrade attack when OpenSSL servers are paired with a number of client behaviors. RFC 8446 makes it unambiguous that those client behaviors are allowed. Not only are they allowed, they are quite realistic given the real-world constraints of post-quantum KEM sizes. That means the fault is on the OpenSSL side.

> Is there an attack here or is this simply a suboptimal outcome?

A downgrade attack is when an attacker can cause you to negotiate a less secure option than you otherwise would have. In this case, if OpenSSL users are concerned about an attacker with a quantum computer, downgrading a connection allows them to mount a store-and-decrypt attack.

So... yes? A suboptimal outcome _is_ the attack here.